hpat/hiframes/api.py

```diff
@@ -1763,3 +1763,68 @@ def _analyze_op_pair_first(self, scope, equiv_set, expr):
 
 
 numba.array_analysis.ArrayAnalysis._analyze_op_pair_first = _analyze_op_pair_first
+
+
+def _append(A, B):
+    return None
+
+
+@overload(_append)
+def _append_overload(A, B):
+    '''Function for appending underlying arrays (A and B) or list/tuple of arrays B to an array A'''
+
+    if isinstance(A, types.Array):
+        if isinstance(B, types.Array):
+            def _append_single_numeric_impl(A, B):
+                return np.concatenate((A, B,))
+
+            return _append_single_numeric_impl
+        elif isinstance(B, (types.UniTuple, types.List)):
+            # TODO: this heavily relies on B being a homogeneous tuple/list - find a better way
+            # to resolve common dtype of heterogeneous sequence of arrays
+            np_dtypes = [numpy_support.as_dtype(A.dtype), numpy_support.as_dtype(B.dtype.dtype)]
+            np_common_dtype = np.find_common_type([], np_dtypes)
+            numba_common_dtype = numpy_support.from_dtype(np_common_dtype)
+
+            # TODO: refactor to use np.concatenate when Numba supports building a tuple at runtime
+            def _append_list_numeric_impl(A, B):
+
+                total_length = len(A) + np.array([len(arr) for arr in B]).sum()
+                new_data = np.empty(total_length, numba_common_dtype)
+
+                stop = len(A)
+                new_data[:stop] = A
+                for arr in B:
+                    start = stop
+                    stop = start + len(arr)
+                    new_data[start:stop] = arr
+                return new_data
+
+            return _append_list_numeric_impl
+
+    elif A == string_array_type:
+        if B == string_array_type:
+            def _append_single_string_array_impl(A, B):
+                total_size = len(A) + len(B)
+                total_chars = num_total_chars(A) + num_total_chars(B)
+                new_data = hpat.str_arr_ext.pre_alloc_string_array(total_size, total_chars)
+
+                list_of_strings = list(A) + list(B)
+                hpat.str_arr_ext.cp_str_list_to_array(new_data, list_of_strings)
+                return new_data
```

Comment:
It doesn't take into consideration `nan`. So appending to string with `nans` will eliminate all `nan` and left empty strings instead (or something worse). I'm not sure if there is an easy way to solve it right now

Reply:
Or am I wrong?

Reply:
@AlexanderKalistratov No, you're right - converting StringArrayType to list transforms nan-s to empty strings, so if None value is added to the test it fails. Thanks! That's definitely a good catch.

```
Series values are different (14.28571 %)
[left]:  [a, , , q, d, , e]
[right]: [a, None, , q, d, , e]
```

Reply:
@AlexanderKalistratov I fixed it by setting null bits while copying all array elements in a loop. It's not as efficient as it could be for the flat arrays, but optimal implementation needs separate low-level copy_to function for string arrays. I think that should be addressed in a different PR.